pdns/dnsdistdist/dnsdist-lbpolicies.cc

```diff
@@ -101,24 +94,56 @@ static shared_ptr<DownstreamState> valrandom(unsigned int val, const ServerPolic
         sum += d.second->d_config.d_weight;
       }
 
-      poss.emplace_back(sum, d.first);
+      poss[usableServers]  = std::make_pair(sum, d.first);
+      usableServers++;
     }
   }
 
   // Catch poss & sum are empty to avoid SIGFPE
-  if (poss.empty() || sum == 0) {
+  if (usableServers == 0 || sum == 0) {
     return shared_ptr<DownstreamState>();
   }
 
   int r = val % sum;
-  auto p = upper_bound(poss.begin(), poss.end(),r, [](int r_, const decltype(poss)::value_type& a) { return  r_ < a.first;});
+  auto p = upper_bound(poss.begin(), poss.begin() + usableServers, r, [](int r_, const typename T::value_type& a) { return  r_ < a.first;});
   if (p == poss.end()) {
```

Comment:
Hmm, this does not feel right. I believe `std::upper_bound` will return `last` of nothing was found. And `last` does not necessarily equals `poss.end()`

Reply:
Ouch, that test is indeed no longer correct, good catch!